pkg/client/client.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return err
+	}
+
+	err = unstructuredToObject(r, obj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *client) Update(ctx context.Context, obj runtime.Object) error {
+	ns, _, err := namespaceAndName(obj)
+	if err != nil {
+		return err
+	}
+	gvks, _, err := c.scheme.ObjectKinds(obj)
+	if err != nil {
+		return err
+	}
+	if len(gvks) == 0 {
+		return fmt.Errorf("object is not registered")
+	}
+	gvk := gvks[0]
+
+	resource, err := apiResource(gvk, c.restMapper)
+	if err != nil {
+		return fmt.Errorf("failed to get resource type: %v", err)
+	}
+
+	result, err := c.dc.Resource(resource).Namespace(ns).Update(unstructuredFromRuntimeObject(obj))
+	if err != nil {
+		return err
+	}
+	err = unstructuredToObject(result, obj)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetNameAndNamespace extracts the name and namespace from the given runtime.Object
+// and returns a error if any of those is missing.
+func namespaceAndName(object runtime.Object) (string, string, error) {
+	accessor := meta.NewAccessor()
+	name, err := accessor.Name(object)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get name for object: %v", err)
+	}
+	namespace, err := accessor.Namespace(object)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get namespace for object: %v", err)
+	}
+	return namespace, name, nil
+}
+
+func (c *client) Delete(ctx context.Context, obj runtime.Object) error {
+	ns, name, err := namespaceAndName(obj)
+	if err != nil {
+		return err
+	}
+	gvks, _, err := c.scheme.ObjectKinds(obj)
+	if err != nil {
+		return err
+	}
+	if len(gvks) == 0 {
+		return fmt.Errorf("object is not registered")
+	}
+
+	gvk := gvks[0]
+
+	resource, err := apiResource(gvk, c.restMapper)
+	if err != nil {
+		return fmt.Errorf("failed to get resource type: %v", err)
+	}
+
+	err = c.dc.Resource(resource).Namespace(ns).Delete(name, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) List(ctx context.Context, namespace string, opts metav1.ListOptions, list runtime.Object) error {
+	gvks, _, err := c.scheme.ObjectKinds(list)
+	if err != nil {
+		return err
+	}
+	if len(gvks) == 0 {
+		return fmt.Errorf("object is not registered")
+	}
+
+	gvk := gvks[0]
+	// TODO(droot): Hack to determine the resource for XXXList kinds.
+	if strings.HasSuffix(gvk.Kind, "List") {
+		gvk.Kind = strings.TrimSuffix(gvk.Kind, "List")
+	}
+
+	resource, err := apiResource(gvk, c.restMapper)
+	if err != nil {
+		return fmt.Errorf("failed to get resource type: %v", err)
+	}
+	fmt.Printf("resource: %v \n", resource)
+
+	ulist, err := c.dc.Resource(resource).Namespace(namespace).List(opts)
+	if err != nil {
+		return err
+	}
+
+	err = unstructuredListToObjectList(ulist, list)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// apiResource consults the REST mapper to translate an <apiVersion, kind, namespace> tuple to a schema.GroupVersionResource struct.
```

Comment:
This comment looks confusing to me. The function accepts `gvk`, I don't see where `namesapce` comes from.